--- dietfeatures.h	2005-10-21 16:14:11.000000000 +0200
+++ dietfeatures.h	2005-10-21 16:15:24.000000000 +0200
@@ -7,7 +7,7 @@
 #define WANT_FLOATING_POINT_IN_SCANF
 #define WANT_CHARACTER_CLASSES_IN_SCANF
 #define WANT_NULL_PRINTF
-/* #define WANT_ERROR_PRINTF */
+#define WANT_ERROR_PRINTF
 #define WANT_LONGLONG_PRINTF
 #define WANT_LONGLONG_SCANF
 
@@ -43,7 +43,7 @@
 /* 20040118: enabling this breaks User Mode Linux!  It's their fault. */
 #define WANT_SYSENTER
 
-#define WANT_LINKER_WARNINGS
+//#define WANT_LINKER_WARNINGS
 
 /* you need to define this if you want to run your programs with large
  * file support on kernel 2.2 or 2.0 */
