--- svscan-add-to-inittab.in	2002-04-23 00:19:38.000000000 +0200
+++ svscan-add-to-inittab.in	2003-05-10 01:38:35.000000000 +0200
@@ -4,10 +4,8 @@
 
 # Look for the svscan lines, and exit if they both exist
 
-SV_pattern='^SV:.*svscan-start'
-SX_pattern='^SX:.*svscan-stopall'
-if egrep "$SV_pattern" "$filename" >/dev/null 2>&1 &&
-   egrep "$SX_pattern" "$filename" >/dev/null 2>&1
+SV_pattern='^SV:.*svscanboot'
+if egrep "$SV_pattern" "$filename" >/dev/null 2>&1
 then
   echo "$0: Nothing to do:"
   echo "  '$filename' appears to already have the svscan lines in it."
@@ -31,7 +29,7 @@
   exit 1
 fi
 
-catit() { cat "$filename" | egrep -v "$SV_pattern" | egrep -v "$SX_pattern"; }
+catit() { cat "$filename" | egrep -v "$SV_pattern"; }
 
 # Insert the two svscan lines between the first rc line and the rest of
 # the inittab so that svscan gets executed before the rc# scripts do.
@@ -54,8 +52,7 @@
 # Splice the new lines into the old file
 {
   catit | head -$(($lineno-1))
-  echo "SV:2345:respawn:$conf_bin/svscan-start /service"
-  echo "SX:S016:wait:$conf_bin/svscan-stopall /service"
+  echo "SV:2345:respawn:$conf_bin/svscanboot"
   catit | tail +$lineno
 } >"$tmpfile"
 
